<a href="https://colab.research.google.com/github/NourFahmy/applications/blob/main/hugging_face_final_DLLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Prompt:
We're going to use the wikitext (link) dataset with the distilbert-base-cased (link) model checkpoint.

Start by loading the wikitext-2-raw-v1 version of that dataset, and take the 11th example (index 10) of the train split.
We'll tokenize this using the appropriate tokenizer, and we'll mask the sixth token (index 5) the sequence.

When using the distilbert-base-cased checkpoint to unmask that (sixth token, index 5) token, what is the most probable predicted token (please provide the decoded token, and not the ID)?

wikitext link: https://huggingface.co/datasets/wikitext/viewer/wikitext-103-raw-v1/train

model link: https://huggingface.co/distilbert-base-cased



In [ ]:
!pip install huggingface_hub
!pip install datasets
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.4/182.4 KB 4.6 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 452.9/452.9 KB 8.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.0/132.0 KB 10.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 213.0/213.0 KB 19.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 140.6/140.6 KB 15.1 MB/s eta 0:00:00
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.8/5.8 MB 53.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 96.

First we import our dataset, tokenizer and model.

In [ ]:
from datasets import load_dataset
from transformers import AutoModel
from transformers import DistilBertTokenizer

dataset = load_dataset("wikitext", "wikitext-2-raw-v1", split="train")
tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')
model = AutoModel.from_pretrained("distilbert-base-cased")

Generating test split:   0%|          | 0/4358 [00:00<?, ? examples/s]

Generating train split:   0%|          | 0/36718 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/3760 [00:00<?, ? examples/s]

Dataset wikitext downloaded and prepared to /root/.cache/huggingface/datasets/wikitext/wikitext-2-raw-v1/1.0.0/a241db52902eaf2c6aa732210bead40c090019a499ceb13bcbfa3f8ab646a126. Subsequent calls will reuse this data.


Downloading:   0%|          | 0.00/232k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/483 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/411 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/263M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert-base-cased were not used when initializing DistilBertModel: ['vocab_projector.bias', 'vocab_layer_norm.weight', 'vocab_transform.bias', 'vocab_projector.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias']
- This IS expected if you are initializing DistilBertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Next we extract the 11th example of the dataset specified in the prompt

In [ ]:
text = dataset[10]['text']
text


' The game \'s battle system , the BliTZ system , is carried over directly from Valkyira Chronicles . During missions , players select each unit using a top @-@ down perspective of the battlefield map : once a character is selected , the player moves the character around the battlefield in third @-@ person . A character can only act once per @-@ turn , but characters can be granted multiple turns at the expense of other characters \' turns . Each character has a field and distance of movement limited by their Action Gauge . Up to nine characters can be assigned to a single mission . During gameplay , characters will call out if something happens to them , such as their health points ( HP ) getting low or being knocked out by enemy attacks . Each character has specific " Potentials " , skills unique to each character . They are divided into " Personal Potential " , which are innate skills that remain unaltered unless otherwise dictated by the story and can either help or impede a charac

Tokenize our specified text using distilbert's tokenizer

In [ ]:
encoded_input = tokenizer(text, return_tensors='pt')
encoded_input

{'input_ids': tensor([[  101,  1996,  2208,  1005,  1055,  2645,  2291,  1010,  1996, 22312,
          2291,  1010,  2003,  3344,  2058,  3495,  2013, 11748,  4801,  7895,
         11906,  1012,  2076,  6416,  1010,  2867,  7276,  2169,  3131,  2478,
          1037,  2327,  1030,  1011,  1030,  2091,  7339,  1997,  1996, 11686,
          4949,  1024,  2320,  1037,  2839,  2003,  3479,  1010,  1996,  2447,
          5829,  1996,  2839,  2105,  1996, 11686,  1999,  2353,  1030,  1011,
          1030,  2711,  1012,  1037,  2839,  2064,  2069,  2552,  2320,  2566,
          1030,  1011,  1030,  2735,  1010,  2021,  3494,  2064,  2022,  4379,
          3674,  4332,  2012,  1996, 10961,  1997,  2060,  3494,  1005,  4332,
          1012,  2169,  2839,  2038,  1037,  2492,  1998,  3292,  1997,  2929,
          3132,  2011,  2037,  2895,  7633,  1012,  2039,  2000,  3157,  3494,
          2064,  2022,  4137,  2000,  1037,  2309,  3260,  1012,  2076, 11247,
          1010,  3494,  2097,  2655,  

Quick check to see what decoded token looks like

In [ ]:
encoded_input = tokenizer(text, return_tensors='pt')
tokenizer.decode(encoded_input["input_ids"][0])

'[CLS] the game\'s battle system, the blitz system, is carried over directly from valkyira chronicles. during missions, players select each unit using a top @ - @ down perspective of the battlefield map : once a character is selected, the player moves the character around the battlefield in third @ - @ person. a character can only act once per @ - @ turn, but characters can be granted multiple turns at the expense of other characters\'turns. each character has a field and distance of movement limited by their action gauge. up to nine characters can be assigned to a single mission. during gameplay, characters will call out if something happens to them, such as their health points ( hp ) getting low or being knocked out by enemy attacks. each character has specific " potentials ", skills unique to each character. they are divided into " personal potential ", which are innate skills that remain unaltered unless otherwise dictated by the story and can either help or impede a character, and

pick the 6th token (5th index)

In [ ]:
for i in range(6):
  print(tokenizer.decode(encoded_input["input_ids"][0][i]))

[ C L S ]
t h e
g a m e
'
s
b a t t l e


this means the first appearance of the word "battle" is our 6th token, we will mask accordingly and modify the string directly
the unmasker takes a string directly, so manually editing the string shouldn't a huge issue especially for our particular use case of one example

In [ ]:
maskedText = '[CLS] the game\'s [MASK] system, the blitz system, is carried over directly from valkyira chronicles. during missions, players select each unit using a top @ - @ down perspective of the battlefield map : once a character is selected, the player moves the character around the battlefield in third @ - @ person. a character can only act once per @ - @ turn, but characters can be granted multiple turns at the expense of other characters\'turns. each character has a field and distance of movement limited by their action gauge. up to nine characters can be assigned to a single mission. during gameplay, characters will call out if something happens to them, such as their health points ( hp ) getting low or being knocked out by enemy attacks. each character has specific " potentials ", skills unique to each character. they are divided into " personal potential ", which are innate skills that remain unaltered unless otherwise dictated by the story and can either help or impede a character, and " battle potentials ", which are grown throughout the game and always grant boons to a character. to learn battle potentials, each character has a unique " masters table ", a grid @ - @ based skill table that can be used to acquire and link different skills. characters also have special abilities that grant them temporary boosts on the battlefield : kurt can activate " direct command " and move around the battlefield without depleting his action point gauge, the character reila can shift into her " valkyria form " and become invincible, while imca can target multiple enemy units with her heavy weapon. [SEP]'

Now that we've tokenized the 11th example and masked the 6th token, we can run it through the unmasker accordingly

In [ ]:
from transformers import pipeline
unmasker = pipeline('fill-mask', model='distilbert-base-uncased')
unmasker(maskedText)

Downloading:   0%|          | 0.00/268M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/466k [00:00<?, ?B/s]

[{'score': 0.17213371396064758,
  'token': 4337,
  'token_str': 'combat',
  'sequence': 'the game\'s combat system, the blitz system, is carried over directly from valkyira chronicles. during missions, players select each unit using a top @ - @ down perspective of the battlefield map : once a character is selected, the player moves the character around the battlefield in third @ - @ person. a character can only act once per @ - @ turn, but characters can be granted multiple turns at the expense of other characters\'turns. each character has a field and distance of movement limited by their action gauge. up to nine characters can be assigned to a single mission. during gameplay, characters will call out if something happens to them, such as their health points ( hp ) getting low or being knocked out by enemy attacks. each character has specific " potentials ", skills unique to each character. they are divided into " personal potential ", which are innate skills that remain unaltered unl

We see the most probable predicted word for the first appearance of "battle" is "combat" with 17%